### PR TITLE
Use sidecar-shim base image with other sidecars and move them under cmd/sidecars

### PR DESCRIPTION
**Pull request by victortoso**
**What this PR does / why we need it**:

This is a follow up from the introduction of [sidecar-shim](https://quay.io/repository/kubevirt/sidecar-shim) https://github.com/kubevirt/kubevirt/pull/9231

With this PR, all hooks are implemented and all sidecars were moved to `cmd/sidecars` 

**Special notes for your reviewer**:

In this interaction we:
1. add support to `PreCloudInitIso` hook as well. Similar to `OnDefineDomain`, the `sidecar-shim` binary will load from `$PATH`  a binary named `preCloudInitIso`. The README.md was updated with more information.

2. move `cmd/example-hook-sidecar` to `cmd/sidecars/smbios` (this one is already using sidecar-shim since https://github.com/kubevirt/kubevirt/commit/d6950fb292b72f3aa9a291a264f76fadfef5fcb9

3. change `example-disk-mutation-hook-sidecar` and `example-cloudinit-hook-sidecar` to use `sidecar-shim` base image and move them under `cmd/sidecars/`  as well

4. move `cmd/network-slirp-binding` sidecar to `cmd/sidecars` without using `sidecar-shim` base image [as preferred](https://github.com/kubevirt/kubevirt/pull/10272#discussion_r1292812049). @ormergi, jfyi about moving it to `cmd/sidecars`

**Release note**:

```release-note
sidecar-shim implements PreCloudInitIso hook
cloudinit and disk-mutation sidecars are now using sidecar-shim base image
```

### CODE REVIEW

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (vasiliy-ul) [inline cmd/sidecars/cloudinit/cloudinit.go:41]: nit: would it be useful to print also the actual error message? wdyt?
```suggestion
		return "", fmt.Errorf("Failed to unmarshal given VMI spec: %v %s", err, vmiJSON)
```
Same with the others `fmt.Errorf(...)` in this file and in `diskimage.go`.

Comment (victortoso) [reply cmd/sidecars/cloudinit/cloudinit.go]: Added this commit https://github.com/kubevirt/kubevirt/pull/10320/commits/8880313e4682f871ca0099517eaef72a26a77097 for `smbios` sidecar because it was missing already this fix from when I moved it to use `sidecar-shim` in d6950fb292b72f3aa9a291a264f76fadfef5fcb9. With both `cloudinit`  and `disk-mutation` sidecars, the diff is very similar  in their *use sidecar-shim* commits. Thanks for the review!

**Review by vasiliy-ul — COMMENTED** [review]
/approve

File: cmd/sidecars/cloudinit/BUILD.bazel
Comment (victortoso) [reply cmd/sidecars/cloudinit/BUILD.bazel:34]: That indeed worked, at least `make bazel-generate` did not complain and `make` seems fine. I'll make this change shortly. Thanks again :)

**Comment by victortoso** [comment]
@0xFelix and @alicefr in case you have the time to take a look :)

**Comment by victortoso** [comment]
Rebased. No real conflicts, just underlying changes brought by https://github.com/kubevirt/kubevirt/pull/10336

**Comment by victortoso** [comment]
/test pull-kubevirt-e2e-k8s-1.26-sig-storage

```
07:43:44:   [FAIL] [sig-storage] [Serial]ImageUpload [storage-req] Upload an image and start a VMI with PVC [test_id:4621] Should succeed [It] PVC [sig-storage, storage-req]
07:43:44:   tests/storage/imageupload.go:179 
```

**Comment by victortoso** [comment]
/test pull-kubevirt-e2e-k8s-1.26-sig-compute
```
08:59:24:   [FAIL] [Serial][sig-compute]VirtualMachineClone Tests VM clone simple VM and cloning operations [It] clone with only some of labels/annotations [sig-compute]
08:59:24:   tests/clone_test.go:183
```

File: cmd/sidecars/smbios/smbios.go
Comment (enp0s3) [inline cmd/sidecars/smbios/smbios.go:48]: I remember I had [issues](https://github.com/kubevirt/kubevirt/pull/3856) with logging the XML, therefore I had to decode it, did the log worked for you in this case?

File: cmd/example-disk-mutation-hook-sidecar/diskimage.go
Comment (enp0s3) [inline cmd/example-disk-mutation-hook-sidecar/diskimage.go]: I think you've committed some temporary code here

File: cmd/sidecars/disk-mutation/diskimage.go
Comment (enp0s3) [inline cmd/sidecars/disk-mutation/diskimage.go:86]: `smbios` => `diskimage`

**Review by enp0s3 — COMMENTED** [review]
@victortoso Hi! Thank you for the effort! I have some small questions below, besides that looks good to me

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (0xFelix) [inline cmd/sidecars/cloudinit/cloudinit.go:39]: Inline check?

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (0xFelix) [inline cmd/sidecars/cloudinit/cloudinit.go:46]: Same.

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (0xFelix) [inline cmd/sidecars/cloudinit/cloudinit.go:50]: Won't this clear all of the provided `UserData`?

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (0xFelix) [reply cmd/sidecars/cloudinit/cloudinit.go:39]: Isn't `vmi` unused otherwise?

File: cmd/sidecars/disk-mutation/diskimage.go
Comment (0xFelix) [inline cmd/sidecars/disk-mutation/diskimage.go:44]: Inline check?

File: cmd/sidecars/disk-mutation/diskimage.go
Comment (0xFelix) [inline cmd/sidecars/disk-mutation/diskimage.go:55]: Same?

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (0xFelix) [reply cmd/sidecars/cloudinit/cloudinit.go:50]: To clarify, this is an example without actual use?

**Review by 0xFelix — COMMENTED** [review]
Added some comments, otherwise I would agree with some of @enp0s3's points.

File: cmd/sidecars/sidecar_shim.go
Comment (victortoso) [reply cmd/sidecars/sidecar_shim.go:64]: Ok!

File: cmd/sidecars/sidecar_shim.go
Comment (victortoso) [reply cmd/sidecars/sidecar_shim.go:67]: Yes, I'll be updating with conditional log, thanks!

File: cmd/sidecars/sidecar_shim.go
Comment (victortoso) [reply cmd/sidecars/sidecar_shim.go:123]: No. I don't think we should be `panic`. I'll change.

File: cmd/sidecars/README.md
Comment (victortoso) [reply cmd/sidecars/README.md:19]: Ok!

File: cmd/sidecars/smbios/smbios.go
Comment (victortoso) [reply cmd/sidecars/smbios/smbios.go:48]: I haven't tested this. I've added the conditional in case of errors, we have something to help debug.
The domainXML should be a valid xml string, no sure what you mean by having to decode it.

File: cmd/sidecars/disk-mutation/diskimage.go
Comment (victortoso) [reply cmd/sidecars/disk-mutation/diskimage.go:86]: Thanks, that was a bad copy paste.

File: cmd/example-disk-mutation-hook-sidecar/diskimage.go
Comment (victortoso) [reply cmd/example-disk-mutation-hook-sidecar/diskimage.go]: I don't know where? Moving to sidecar-shim container image means using pflag for the binary to receive the vmi (in json format) and the domain (in xml format). I might be missing it, so let me know

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (victortoso) [reply cmd/sidecars/cloudinit/cloudinit.go:39]: Indeed! This is true for the [current code](https://github.com/kubevirt/kubevirt/blob/main/cmd/example-cloudinit-hook-sidecar/cloudinit.go#L73). This example sidecar is not using it. I'll remove.

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (victortoso) [reply cmd/sidecars/cloudinit/cloudinit.go:50]: That's correct, just an example we have. My intentional is to move all sidecars to the same folder and use `sidecar-shim` where it is possible. I don't mind misc fixes around them. Thanks for the review!

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (victortoso) [reply cmd/sidecars/cloudinit/cloudinit.go:46]: I usually love `inline` check but some people doesn't.
In this case, I didn't change this code, only moved it.
I don't mind changing it if you feel strong about this.

File: cmd/sidecars/disk-mutation/diskimage.go
Comment (victortoso) [reply cmd/sidecars/disk-mutation/diskimage.go:44]: Again, this is how the code is [at the present](https://github.com/kubevirt/kubevirt/blob/main/cmd/example-disk-mutation-hook-sidecar/diskimage.go#L74), I've only applied the changes related to sidecar-shim plus moving to `cmd/sidecar` folder. I don't mind doing this changes as follow up for example.

File: cmd/sidecars/disk-mutation/diskimage.go
Comment (victortoso) [reply cmd/sidecars/disk-mutation/diskimage.go:55]: [ditto](https://github.com/kubevirt/kubevirt/blob/main/cmd/example-disk-mutation-hook-sidecar/diskimage.go#L90)

**Comment by victortoso** [comment]
@0xFelix, @enp0s3 sorry for taking so long to get back to you and many thanks for the reviews!

I've rebased, that might be annoying to see the diff so I've pushed the extra patches into this [use-sidecar-shim-v2](https://github.com/victortoso/kubevirt/commits/use-sidecar-shim-v2) branch. I've squashed one commit as suggested in https://github.com/kubevirt/kubevirt/pull/10320#discussion_r1313773813, one fixup patch and all the others are in the end of the commit history.

I'm making taken suggestions as `resolved`. Let me know if I got something wrong. Cheers!

File: cmd/sidecars/cloudinit/cloudinit.go
Comment (victortoso) [reply cmd/sidecars/cloudinit/cloudinit.go:39]: Actually, I left it for now. I'm not sure if the intention was to have it show in the example. I can do a follow up, perhaps using some `vmi` data into `cloudInitData` just to make usage of vmi... or properly remove it.

**Comment by victortoso** [comment]
/retest-required

```
22:07:04: Summarizing 1 Failure:
22:07:04:   [FAIL] [sig-compute][Serial]CPU Hotplug A VM with cpu.maxSockets set higher than cpu.sockets [It] should successfully plug guaranteed vCPUs [sig-compute, sig-compute-migrations, requires-two-schedulable-nodes, VMLiveUpdateFeaturesGate]
22:07:04:   tests/hotplug/cpu.go:277
22:07:04: 
22:07:04: Ran 40 of 1532 Specs in 3285.425 seconds
22:07:04: FAIL! -- 39 Passed | 1 Failed | 2 Pending | 1490 Skipped
22:07:04: --- FAIL: TestTests (3285.75s)
22:07:04: FAIL
```

**Comment by alicefr** [comment]
/lgtm
/approve
/hold

Looks good to me, let give @enp0s3 and @0xFelix some time for a last look

**Review by 0xFelix — APPROVED** [review]
Looks good to me, thanks!

/lgtm

**Comment by victortoso** [comment]
@enp0s3 is it fine for you? can we unhold this one? :)

**Comment by victortoso** [comment]
Rebasing due conflicts in `hack/config-default.sh`, came from ab47865130d8d40f56831ed51b2bd3e4d8989ca0 (add's `virt-tail` in binaries)

```
<<<<<<< use-sidecar-shim
binaries_and_docker_images="
    cmd/virt-operator
    cmd/virt-controller
    cmd/virt-launcher
    cmd/virt-handler
    cmd/virt-api
    cmd/sidecars
    cmd/sidecars/smbios
    cmd/sidecars/cloudinit
"

binaries="
    ${binaries_and_docker_images}
    cmd/virtctl
    cmd/fake-qemu-process
    cmd/virt-chroot
"

docker_images="
    ${binaries_and_docker_images}
    images/disks-images-provider
    images/vm-killer
    images/nfs-server
    images/winrmcli
    tests/conformance
"

=======
binaries="cmd/virt-operator cmd/virt-controller cmd/virt-launcher cmd/virt-handler cmd/virtctl cmd/fake-qemu-process cmd/virt-api cmd/example-hook-sidecar cmd/example-cloudinit-hook-sidecar cmd/virt-chroot cmd/sidecars cmd/virt-tail"
docker_images="cmd/virt-operator cmd/virt-controller cmd/virt-launcher cmd/virt-handler cmd/virt-api images/disks-images-provider images/vm-killer images/nfs-server images/winrmcli cmd/example-hook-sidecar cmd/example-cloudinit-hook-sidecar tests/conformance cmd/sidecars"
>>>>>>> main
```

**Review by 0xFelix — APPROVED** [review]
/lgtm

Let's get this merged

**Comment by victortoso** [comment]
> Let's get this merged

Yep, @0xFelix we have 2 approves and 1 lgtm, should be fine to unhold it I hope? :)

**Comment by 0xFelix** [comment]
Yes, this no longer needs to be on hold!

/hold cancel

**Pull request merged**
